components/javascript-wrapper.js

```diff
@@ -0,0 +1,157 @@
+var rpf = require('./rpf');
+
+/**
+ * This module provides a Javascript wrapper for updaters.  It is responsible for setting up the
+ * code to ensure updater callback, and then invoking the RDF pipeline framework to complete building
+ * the component and setting up the vni and state metadata.
+ *
+ * @param nodeDef updater node definition. This can be a simple updater function, or an object that contains 
+ *        updater metadata (e.g., input/output port definitions, description, icons), and an updater.
+ *
+ * @return a promise to create the noflo rdf component
+ */
+module.exports = function(nodeDef) { 
+
+    // Save the caller's node instance 
+    var self = this;
+    
+    // Ensure we have at least one input port and and updater, using a default stub updater if 
+    // there is none specified
+    var updaterArgs = null;
+    if ( _.isUndefined(nodeDef) || 
+       ( ! _.isFunction(nodeDef) && ( _.isEmpty( nodeDef ) || ( _.isEmpty( nodeDef.inPorts ))))) {
+       // Got nothing to work with - no input ports and no updater, so create a default updater and input port 
+       nodeDef = { 
+           inPorts: {
+               input: {
+                   datatype: 'object',
+                   description: "default input port" 
+               }
+           },
+           updater: defaultUpdater
+       };
+
+    }  else if ( _.isUndefined( nodeDef.updater ) && !_.isFunction(nodeDef)) { 
+       // Got some node metadata, but no node updater - use the metadata with default updater
+       nodeDef = _.extend( nodeDef, 
+                           { updater: defaultUpdater } );
+
+    } else if ( _.isFunction( nodeDef ) ) { 
+        // Got an updater but no metadata - use updater args as input port list
+        // This assumes that all updater arguments are input
+        updaterArgs = introspect( nodeDef );
+        nodeDef = { inPorts: updaterArgs,
+                    updater: nodeDef };
+    } 
+
+    // Have an updater now, no matter what, be it default or not - figure out what the updater arguments are
+    var updaterArgs = (_.isNull( updaterArgs )) ? introspect( nodeDef.updater ) : updaterArgs; 
+    var inPorts = _.isArray(nodeDef.inPorts) ? _.object(nodeDef.inPorts, []) : nodeDef.inPorts;
+
+    return promiseComponent( 
+                         _.defaults({
+                             inPorts: _.mapObject(inPorts, function(port, name) {
+                                 return _.extend( port, 
+                                                  { ondata: ondata.bind( self, name, updaterArgs, nodeDef )});
+                             })
+                           }, nodeDef)
+                     );
+
+}; // module.exports
+
+// Define a default updater stub function to be used if the caller 
+// did not pass one in.  Simply returns whatever the original input was so it
+// can be sent through the output port to the next component
+var defaultUpdater = function() { 
+    return { out: arguments };
```

Comment:
Excellent.  However, since we changed to allow only one state per node, the updater API can be simplified to just return the new output state data, instead of including the port name.  For example, if there is only one updater parameter 'input', then an updater could be written:
<pre>
function myUpdater(input) { return input; }
</pre>

We didn't discuss what inputs the defaultJavascriptUpdater should have, or what it should do. How about if we do the following:

 <pre>
1. Automatically add an inPort for every updater parameter 
(except 'output') that is not already listed in inPorts.
2. if (no updater and no inPorts) { add port 'input' to inPorts }
3. if (no updater) { 
     Make a default updater, using the inPort names as parameters.
     If there is only one inPort, have it return that data directly.
     Otherwise, have it return an object whose keys are the parameter
     (inport) names, and whose values are the values of those 
     parameters.  For example, if the updater has two parameters in1
     and in2, then it would return: { in1: in1, in2: in2 }
}
</pre>